primitives/session/src/lib.rs

```diff
@@ -106,6 +106,23 @@ impl GetValidatorCount for MembershipProof {
 	}
 }
 
+/// Trait for retrieving the session info needed for online node inspection.
+///
+/// This trait is used for decouple the pallet-session dependency from im-online
+/// module so that the user of im-online & offences modules can pass any list of
+/// validators that are considered to be online in each session, particularly useful
+/// for the Substrate-based projects having their own staking implementation
+/// instead of using pallet-staking directly.
+pub trait SessionInterface<ValidatorId> {
+	/// Returns current session index.
+	fn current_index() -> SessionIndex;
+
+	/// Returns all the validators ought to be online in a session.
+	///
+	/// The returned validators are all expected to be running an authority node.
+	fn validators() -> Vec<ValidatorId>;
```

Comment:
The fact that we still have to rely on `pallet_session` for offences reporting makes it clear that the decoupling is not complete yet. IMHO the abstraction is a bit leaking now, IMHO a `ValidatorSet` abstraction with offence reporting capabilities (extend `ReportOffence` trait) should be the way to go and will allow us to get rid of `pallet_session` dependency.